drivers/counter/ds3231.c

```diff
@@ -0,0 +1,1060 @@
+	}
+	if (data->isw == NULL) {
+		return -ENOTSUP;
+	}
+	if (data->sync_state != SYNCSM_IDLE) {
+		return -EBUSY;
+	}
+
+	data->sync_signal = signal;
+	data->sync_state = SYNCSM_PREP_READ;
+	k_work_submit(&data->sync_work);
+
+	return 0;
+}
+
+static int ds3231_set(struct device *dev,
+		      const struct rtc_ds3231_syncpoint *syncpoint,
+		      struct k_poll_signal *signal)
+{
+	struct ds3231_data *data = dev->driver_data;
+
+	if ((syncpoint == NULL)
+	    || (signal == NULL)) {
+		return -EINVAL;
+	}
+	if (data->isw == NULL) {
+		return -ENOTSUP;
+	}
+	if (data->sync_state != SYNCSM_IDLE) {
+		return -EBUSY;
+	}
+
+	data->new_sp = *syncpoint;
+	data->sync_signal = signal;
+	data->sync_state = SYNCSM_PREP_WRITE;
+	k_work_submit(&data->sync_work);
+
+	return 0;
+}
+
+
+static int ds3231_init(struct device *dev)
+{
+	struct ds3231_data *data = dev->driver_data;
+	const struct ds3231_config *cfg = dev->config->config_info;
+	struct device *i2c = device_get_binding(cfg->bus_name);
+	int rc;
+
+	data->ds3231 = dev;
+	if (i2c == NULL) {
+		LOG_WRN("Failed to get I2C %s", cfg->bus_name);
+		return -EINVAL;
+	}
+
+	data->i2c = i2c;
+	rc = update_registers(dev);
+	if (rc < 0) {
+		LOG_WRN("Failed to fetch registers: %d", rc);
+		return rc;
+	}
+
+	/* INTCN and AxIE to power-up default, RS to 1 Hz */
+	rc = sc_ctrl(dev,
+		     RTC_DS3231_REG_CTRL_INTCN,
+		     RTC_DS3231_REG_CTRL_RS_Msk | RTC_DS3231_ALARM1 | RTC_DS3231_ALARM2);
+	if (rc < 0) {
+		LOG_WRN("Failed to reset config: %d", rc);
+		return rc;
+	}
+
+	if (cfg->isw_gpios.ctrl != NULL) {
+		struct device *gpio = device_get_binding(cfg->isw_gpios.ctrl);
+
+		if (gpio == NULL) {
+			LOG_WRN("Failed to get INTn/SQW GPIO %s", cfg->isw_gpios.ctrl);
+			return -EINVAL;
+		}
+
+		k_timer_init(&data->sync_timer, sync_timer_handler, NULL);
+		k_work_init(&data->alarm_work, alarm_worker);
+		k_work_init(&data->sqw_work, sqw_worker);
+		k_work_init(&data->sync_work, sync_worker);
+		gpio_init_callback(&data->isw_callback,
+				   isw_gpio_callback,
+				   BIT(cfg->isw_gpios.pin));
+
+		rc = gpio_pin_configure(gpio, cfg->isw_gpios.pin,
+					GPIO_DIR_IN
+					| GPIO_PUD_PULL_UP
+					| GPIO_INT
+					| GPIO_INT_EDGE
+					| GPIO_INT_ACTIVE_LOW
+					| GPIO_INT_DEBOUNCE);
+		if (rc >= 0) {
+			rc = gpio_add_callback(gpio, &data->isw_callback);
+		}
+		if (rc >= 0) {
+			data->isw = gpio;
+		} else {
+			LOG_WRN("Failed to configure ISW callback: %d", rc);
+		}
+	}
+
+	LOG_DBG("Initialized");
+	return 0;
+}
+
+static int ds3231_counter_start(struct device *dev)
+{
+	return -EALREADY;
+}
+
+static int ds3231_counter_stop(struct device *dev)
+{
+	return -ENOTSUP;
+}
+
+int ds3231_counter_set_alarm(struct device *dev,
+			     u8_t id,
+			     const struct counter_alarm_cfg *alarm_cfg)
+{
+	struct ds3231_data *data = dev->driver_data;
+	const struct register_map *rp = &data->registers;
+	int rc = 0;
+
+	if (rp->ctrl & (RTC_DS3231_ALARM1 << id)) {
+		return -EBUSY;
+	}
+
+	time_t when = alarm_cfg->ticks;
```

Comment:
this assignment is redundant in case of absolute. What about moving it to else clause (if not absolute)?

Reply:
done